internal/characters/lyney/burst.go

```diff
@@ -0,0 +1,142 @@
+package lyney
+
+import (
+	"github.com/genshinsim/gcsim/internal/frames"
+	"github.com/genshinsim/gcsim/pkg/core/action"
+	"github.com/genshinsim/gcsim/pkg/core/attacks"
+	"github.com/genshinsim/gcsim/pkg/core/attributes"
+	"github.com/genshinsim/gcsim/pkg/core/combat"
+	"github.com/genshinsim/gcsim/pkg/core/event"
+	"github.com/genshinsim/gcsim/pkg/enemy"
+)
+
+// 3 * 60 duration
+// burst, none, default, default, no hitlag
+// every 0.15*60 perform true ST attack with burst
+// at end of duration: explosiveFirework with 6m aoe around player
+// burst ends either via duration or swap
+// 15s cooldown
+// energy drain
+
+var burstFrames []int
+
+const (
+	// TODO: proper frames
+	burstKey         = "lyney-q"
+	burstMarkKey     = "lyney-burst-mark"
+	burstInterval    = 0.15 * 60
+	burstDuration    = 3*60 + 1 // + 1 for final tick
+	burstCD          = 15 * 60
+	burstCDStart     = 1
+	burstEnergyDelay = 7
+)
+
+func init() {
+	// TODO: proper min frames, currently using thoma
+	burstFrames = frames.InitAbilSlice(58)
+	burstFrames[action.ActionAttack] = 57
+	burstFrames[action.ActionSkill] = 56
+	burstFrames[action.ActionDash] = 57
+	burstFrames[action.ActionSwap] = 56
+}
+
+// Burst attack damage queue generator
+func (c *char) Burst(p map[string]int) action.ActionInfo {
+	c.AddStatus(burstKey, burstDuration, true)
+	c.QueueCharTask(c.burstTick, burstInterval)
+	c.QueueCharTask(c.explosiveFirework, burstDuration)
+
+	c.SetCDWithDelay(action.ActionBurst, burstCD, burstCDStart)
+	c.ConsumeEnergy(burstEnergyDelay)
+
+	return action.ActionInfo{
+		Frames:          frames.NewAbilFunc(burstFrames),
+		AnimationLength: burstFrames[action.InvalidAction],
+		CanQueueAfter:   burstFrames[action.ActionSkill], // TODO: proper frames, should be earliest cancel
+		State:           action.BurstState,
+	}
+}
+
+func (c *char) onExitField() {
+	c.Core.Events.Subscribe(event.OnCharacterSwap, func(args ...interface{}) bool {
+		prev := args[0].(int)
+		if prev != c.Index {
+			return false
+		}
+		if !c.StatusIsActive(burstKey) {
+			return false
+		}
+		c.explosiveFirework()
+		return false
+	}, "lyney-exit")
+}
```

Comment:
Canceling Q is not possible via Swap. It's only possible via E while Q is active, E should not do anything except cancel Q state if done during Q.